EZP-24615: Improve semantic configuration and update documentation

clash82 opened the pull request:
Link to Jira task: https://jira.ez.no/browse/EZP-24615

- [x] fixed semantic configuration issues (no changes to user part, it's fully compatible with alfa3)
- [x] updated README: added arguments naming in `yc_show_recommendations()` Twig helper (version without argument naming is also supported); semantic configuration improvement (changed advanced settings example); raised requirements for eZ Publish / Platform

bdunogier commented:
I'd describe the actual config changes in the PR's description.

bdunogier commented on README.md:
```diff
@@ -104,30 +115,40 @@ Implementation is very easy and can be performed in just a few steps (assuming t
+`contentType`   | string | content type values you are expecting in response
 `template`      | string | HandleBars template name (your templates are stored under `EzRecommendationBundle/Resources/public/views` directory. Take a look on `default.html.twig` file which includes default template that can be used to prepare customised version)
-`[fields]`      | array  | here you can define which fields are required and will be requested from the recommender engine. These field names are also used inside HandleBars templates
+`attributes`    | array  | here you can define which fields are required and will be requested from the recommender engine. These field names are also used inside HandleBars templates
```
Why did you change `fields` to `attributes` ? The latter isn't part of our Domain Language, while field is.

clash82 replied:
This is the attribute's keyword, it can be omitted.

clash82 replied:
Hmm, yes it may be a good idea to change it :/

clash82 commented:
@bdunogier no important changes to existing configuration (changed notation for advanced settings). I fixed configuration validation and added support for arguments naming in Twig helper + documentation updates.

lolautruche commented on Resources/config/default_settings.yml:
```diff
@@ -1,11 +1,7 @@
+    consume_timeout: 20
+tracking:
+    api_endpoint: 'http://event.yoochoose.net'
+    script_url: 'cdn.yoochoose.net/yct.js'
```
`default_settings.yml` is not supposed to store semantic config, but *internal* default values (like what was in the file before, e.g. `ez_recommendation.default.server_uri`), that would have resulted by the semantic config processing.
Check [what we have in ezpublish-kernel](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Bundle/EzPublishCoreBundle/Resources/config/default_settings.yml) for instance.

This is done to ensure we have correct default values.

lolautruche replied:
Alternative is just to rename this file to something else :)

clash82 replied:
Changed :)

andrerom commented on README.md:
```diff
@@ -104,30 +115,40 @@ Implementation is very easy and can be performed in just a few steps (assuming t
 
 Parameter       | Type   | Description
 ----------------|--------|------------
 `content.id`    | int    | this is in content based views normally the twig variable holding the content id (we want to get recommendations for)
```
I guess the parameter name is not really "content.id"

clash82 replied:
Parameter name is `contentId` but because this part is always required and never changed I decided to leave it as it is.

andrerom replied:
yes but the name of the parameter is wrong. this is mixing parameter name with the name of the variable used in the example. please fix.

andrerom commented:
Looks like good direction, but I'm no config expert so I'll let @bdunogier @michalpipa and @lolautruche give input on that.

But would indeed be nice to have description of changes you are doing in PR's, task list is highly recommended to make it clear what you intend to do, and if you are done or not. Example:

- [x] Change `included_content_types` to scalar for reason X
- [x] Moved yoochoose specific settings to root as they don't need to be configured by siteaccess
- [ ] Fix `included_content_types` seemingly not being required anymore after reformating

:)

clash82 commented:
@andrerom Updated, thank you ;)

Waiting for review.

andrerom commented on DependencyInjection/Configuration.php:
```diff
@@ -30,34 +30,10 @@ public function getConfigTreeBuilder()
-                        ->info('YooChoose tracking script address')
-                        ->example('cdn.yoochoose.net/yct.js')
+                        ->example(array('article', 'blog_post'))
+                        ->prototype('scalar')->end()
```
why does this have a `end()`

clash82 replied:
because it closes `prototype()` not `arrayNode()`

andrerom commented on README.md:
```diff
@@ -104,30 +111,40 @@ Implementation is very easy and can be performed in just a few steps (assuming t
 * place dedicated Twig helper in place where you want to display recommendations:
```
```diff
- place dedicated Twig helper in place where you want to display recommendations:
+ place dedicated Twig helper in place where you want to display recommendations *(see further below for example)*:

lolautruche commented on Resources/config/default_settings.yml:
```diff
@@ -1,9 +1,4 @@
 parameters:
-    ez_recommendation.default.yoochoose.customer_id: ~
-    ez_recommendation.default.yoochoose.license_key: ~
```
Are those not needed any more ?

clash82 replied:
Not in the `default_settings.yml` anymore, it should be defined in `config.yml` by user.

andrerom commented:
+1 (but I'm also curious if the settings @lolautruche comment on as well)

clash82 commented:
> Also, aren't any of the settings site access aware at all? Not something to fix in this PR necessarily, but in a followup that would be nice to clarify.

User settings (customer_id, license_key, server_uri and included_content_types) - they should be siteaccess aware. I will take a closer look later.

andrerom commented:
@clash82 But I guess this PR is still ready for merge right? easier to review if the siteaccess problem is handled as separate topic.

clash82 commented:
@andrerom yes, it's ready for merge (after rebase).

I have replied to @lolautruche about why I decided to remove those settings https://github.com/ezsystems/EzSystemsRecommendationBundle/pull/16#discussion_r35219901 ...did I miss something? :)

lolautruche commented:
+1

andrerom commented:
> did I miss something?

not anymore, ready for rebase then :)

clash82 commented:
Done :)

michalpipa commented:
:+1:

andrerom commented:
merged :)

clash82 commented:
Thank you :)
